optaplanner-website-root/content/learn/drl-to-constraint-streams-migration.adoc

```diff
@@ -0,0 +1,447 @@
+
+## Benefits of Constraint Streams
+
+CS is a modern full-featured API for writing OptaPlanner constraints, offering several benefits over score DRL:
+
+* Developers do not need to learn any new language, CS is plain Java.
+* Full IDE support incl. syntax highlighting and refactoring.
+* Comprehensive support for unit testing.
+* In certain use cases, CS performs considerably faster than score DRL.
+
+## Migrating to Constraint Streams, step by step
+
+CS and DRL are very similar in their approach and in fact both use Drools to execute the constraint logic.
+Because of this, migrating many parts of score DRL to CS is straight-forward, even mechanical.
+However, this guide will not help you if your score DRL uses the following constructs, as there is no direct mapping between DRL and CS in that case:
+
+* DRL's `insertLogical()` allows information to be transferred between rules.
+This is not possible in CS, where each constraint is isolated and stands on its own.
+* DRL's `eval()` allows for arbitrary Java code execution anywhere on the left-hand side of a rule.
+This is not possible in CS.
+* DRL's right-hand side allows for arbitrary Java code execution.
+This is not possible in CS, where each constraint can only result in either a score reward or a penalty.
+* Using more than 3 joins in a single rule, which is very unlikely.
+
+If you are using any of these constructs, we recommend that you first refactor them out of your DRL and then check back with this migration guide.
+
+### The basics
+
+In score DRL, all your constraints are typically written in a single text file like so:
+
+[source, drl]
+----
+package org.optaplanner.examples.vehiclerouting.optional.score;
+    dialect "java"
+
+import ...;
+
+global HardSoftLongScoreHolder scoreHolder;
+
+rule "vehicleCapacity"
+    when
+        ...
+    then
+        ...
+end
+
+...
+----
+
+In this file, each rule represents 1 or more constraints.
+In CS, the DRL file is replaced by a standard Java source code:
+
+[source, java]
+----
+package org.optaplanner.examples.vehiclerouting.score;
+
+import ...;
+
+public class VehicleRoutingConstraintProvider implements ConstraintProvider {
+
+    @Override
+    public Constraint[] defineConstraints(ConstraintFactory factory) {
+        return new Constraint[] {
+                vehicleCapacity(factory),
+                ...
+        };
+    }
+
+    protected Constraint vehicleCapacity(ConstraintFactory factory) {
+        ...
+    }
+
+    ...
+}
+
+----
+
+The method `defineConstraints(...)`,  a single method on the `ConstraintProvider`, lists all your constraints.
+Each constraint is then typically represented by a method.
+
+#### Solver configuration
+
+Users of Quarkus most likely do not need to worry about solver configuration.
+Removing score DRL and introducing a `ConstraintProvider` implementation ought to be enough.
```

Comment:
- Writing guidelines: do not use "maybe", "ought to be", "probably", etc.
- active tense: removing -> rmove
```suggestion
Just remove the score DRL and create a `ConstraintProvider` implementation.
```